rfcs/react-components/convergence/slot-render-as.md

```diff
@@ -0,0 +1,155 @@
+# RFC: Use `as` instead of `children` for slot render function
+
+Contributors: behowell
+
+## Summary
+
+This proposes modifying the slot render function API to pass a render function via the slot's `as` prop instead of the `children` prop. This has several benefits, the primary of which is that it allows the user to write a render function for a slot without overriding the default children of a slot.
+
+This is a complementary proposal to bsunderhus's [RFC #27164: slot children render function support](https://github.com/microsoft/fluentui/pull/27164). It provides an alternative way to avoid having a render function override a slot's default `children` prop. This RFC has no dependencies on it, and could be adopted separately if needed.
+
+## Background
+
+Slots accept render functions that allow them to override the component and/or props used when rendering the slot. For more background, see https://react.fluentui.dev/?path=/docs/concepts-developer-customizing-components-with-slots--page#replacing-the-entire-slot
+
+The render function is passed in as the slot's `children` prop.
+
+## Problem statement
+
+A few problems with render functions as they are today:
+
+1. Using `children` as the render function will override any default children provided by the component's internals.
+2. It is confusing that passing a render function as `children` would override the _whole slot_, and not just its children.
+3. The root slot's render function conflicts the use of children as a function, which certain components use for custom rendering of their children.
+   - For example, Tooltip and Field both accept a function as a child of their root, in order to pass props to the children. However, this breaks the ability to use a slot render function for the root slot of those controls.
+   - See: https://react.fluentui.dev/?path=/docs/preview-components-field--default#complex-content-in-a-field
+
+And a less important issue that's addressed in Appendix A of this proposal:
+
+4. The order of the arguments is `(Component, props)`, but one of the primary use cases is to replace the given slot component with something else, in which case you ignore the first argument. You'd rarely/never(?) want to ignore the second argument `props`.
```

Comment:
I would that it's also opinionated and depends on the app. I checked first 10 usages of `children()` in Teams:

- 6 for `(Component, props)
- 2 for `(_, props)`
- 2 for `()`

Reply:
👆 FWIW I'd call that data in favor of my recommendation 🙂. Specifically, there are _zero_ examples of `(Component, _)`. So it would be better to swap the order of params so that the `(_, props)` cases can be simpler.

Also, one more motivation is that since our `as` prop _does not_ support `as={Component}`, it makes it slightly simpler to write: `as={props => <Component {...props} />}`.

Reply:
We'll remove Appendix A from the proposal, per https://github.com/microsoft/fluentui/pull/27353#discussion_r1152436903. So, resolving this conversation.